infra/blueprint-test/modules/tft/terraform.go

```diff
@@ -0,0 +1,248 @@
+type TFBlueprintTest struct {
+	name        string                 // descriptive name for the test
+	fixtureName string                 // fixture name which will be used to compute TFDir
+	tfDir       string                 // directory containing Terraform configs
+	envVars     map[string]string      // variables to pass to Terraform as environment variables
+	setupPath   string                 // optional directory containing applied TF configs to import outputs as variables for the test
+	vars        map[string]interface{} // variables to pass to Terraform as flags
+	logger      *logger.Logger         // custom logger
+	t           testing.TB             // TestingT or TestingB
+}
+
+type option func(*TFBlueprintTest)
+
+func WithName(name string) option {
+	return func(f *TFBlueprintTest) {
+		f.name = name
+	}
+}
+
+func WithFixtureName(fixtureName string) option {
+	return func(f *TFBlueprintTest) {
+		f.fixtureName = fixtureName
+	}
+}
+
+func WithTFDir(tfDir string) option {
+	return func(f *TFBlueprintTest) {
+		f.tfDir = tfDir
+	}
+}
+
+func WithEnvVars(envVars map[string]string) option {
+	return func(f *TFBlueprintTest) {
+		f.envVars = envVars
+	}
+}
+
+func WithSetupPath(setupPath string) option {
+	return func(f *TFBlueprintTest) {
+		f.setupPath = setupPath
+	}
+}
+
+func WithVars(vars map[string]interface{}) option {
+	return func(f *TFBlueprintTest) {
+		f.vars = vars
+	}
+}
+
+func WithLogger(logger *logger.Logger) option {
+	return func(f *TFBlueprintTest) {
+		f.logger = logger
+	}
+}
+
+// Init sets defaults, validates and returns a TFBlueprintTest.
+func Init(t testing.TB, opts ...option) *TFBlueprintTest {
+	tft := &TFBlueprintTest{
+		name: fmt.Sprintf("%s TF Blueprint", t.Name()),
+		t:    t,
+	}
+	// apply options
+	for _, opt := range opts {
+		opt(tft)
+	}
+	// if no custom logger, set default based on test verbosity
+	if tft.logger == nil {
+		tft.logger = utils.GetLoggerFromT()
+	}
+	// one of fixture name or tfDir should be provided
+	if tft.fixtureName != "" && tft.tfDir != "" {
+		t.Fatalf("Only one of FixtureName or TFDir must be provided, found FixtureName=%s, TFDir=%s", tft.fixtureName, tft.tfDir)
+	}
+	// if both fixture name and tfDir are empty, auto discover tfDir based on cwd
+	if tft.fixtureName == "" && tft.tfDir == "" {
+		cwd, err := os.Getwd()
+		if err != nil {
+			t.Fatalf("unable to get wd :%v", err)
+		}
+		tfdir, err := discovery.ConfigDirFromCWD(cwd)
+		if err != nil {
+			t.Fatalf("unable to detect TFDir :%v", err)
+		}
+		tft.tfDir = tfdir
+	}
+	// if fixture name compute tfDir from given fixture name
+	if tft.fixtureName != "" {
+		tfModFixtureDir := getTFModuleFixtureDir(tft.fixtureName)
+		if _, err := os.Stat(tfModFixtureDir); os.IsNotExist(err) {
+			t.Fatalf("TFDir path derived from %s as %s does not exist", tft.fixtureName, tfModFixtureDir)
+		}
+		tft.tfDir = tfModFixtureDir
+	}
```

Comment:
I suggest moving this part into a separate function as it all deals with computing tfDir and you can exit the function early.

Reply:
I did not this move into a separate function as this had been reduced to just a check if tfDir is explicitly specified else auto discovery. In both cases for errors we fail early via `t.Fatal`. I felt the current approach was readable, but happy to split into a function if you think otherwise.